app/models/post.rb

```diff
@@ -1129,7 +1131,13 @@ def cannot_permanently_delete_reason(user)
   end
 
   def mentions
-    PrettyText.extract_mentions(Nokogiri::HTML5.fragment(cooked))
+    if Discourse.redis.hexists("post_mentions", id)
+      Discourse.redis.hget("post_mentions", id).split(",")
+    else
+      extracted_mentions = PrettyText.extract_mentions(Nokogiri::HTML5.fragment(cooked))
+      Discourse.redis.hset("post_mentions", id, extracted_mentions.join(","))
```

Comment:
I'm not sure if storing the username is ideal here because a user's username can be changed. I'm thinking that storing the user's id might be better.

Reply:
I agree that using id would be better, but we have only username at hand here. We'd need to query the database to get user ids and to take care of possible N + 1 queries then. And all that only to handle a username change, which is a very rare occasion.

In general, this should have been working correctly with changed usernames. If a user changes username:
1. We [recook](https://github.com/discourse/discourse/blob/dfc449a32e9c1fda375e104510c7b28a8171e4c1/app/jobs/regular/update_username.rb#L119) posts with mentions of him, and 
2. We invalidate mentions cache when recooking. 

One thing though is that I double-checked and turned out that recooking after changing username happens in another place, so cache invalidation wasn't working in this case. I pushed a commit that fixes it.

Making query to get ids seems like overkill to me, honestly, I don't see what else can go wrong here.

Reply:
> We'd need to query the database to get user ids and to take care of possible N + 1 queries then.

I'm not sure I understand the problem here. Querying the database only happens when we're setting the cache and the query is a very cheap operation. We're also just plucking the `ids` from the database so I'm not quite sure why we might have to deal with the possibility of an N+1? 

Personally I'll take the extra query when setting the cache then having to deal with invalidating the cache when a username has been updated because the former is much more straight forward than the latter.

Reply:
> We're also just plucking the `ids` from the database so I'm not quite sure why we might have to deal with the possibility of an N+1?

Yeah, we'll just pluck ids for a post. But when we're loading a topic there are many posts, and if the cache is empty we'll be plucking user ids for every post with mentions in the topic. Maybe those queries won't add a lot to query time (but they will add something for sure, while with usernames we only do parsing and no additional queries to db). I'm checking how much those queries will add.

Reply:
But since all the posts have been already loaded and mentions parsed. We can combine the lookup for all usernames into a single query?

Reply:
We can't always combine, sometimes we have several posts, sometimes we have only one (for example when you post and the post is added on the bottom of the topic), we have to handle those cases separately. It's all makes it more complex, usernames are easier, no additional queries, no need to handle different cases separately.

Reply:
Using Ids:
- adds more code
- increases loading time (though probably not significantly)

Usernames already works, there is the problem with renaming, but it's already handled with one line of code.

Reply:
Sorry, probably this will be easier to follow - when we have a topic with several posts we handle it and combine queries when needed in [TopicView](https://github.com/discourse/discourse/blob/main/lib/topic_view.rb). When we have only one post whe don't have topic view, we have only a post model.

So if we want load ids I'll need to add code both to the post model (to handle case when there is no TopicView at hand), and to TopicView (to combine queries).

Right now, everithyng is hadled in the post model, since we don't need to query db and to combine queries. It's simpler.

Reply:
> So if we want to load ids I'll need to add code both to the post model (to handle the case when there is no TopicView at hand), and to TopicView (to combine queries).

I'm not sure if this is necessary. `Post#mentions` returns usernames right now. Even though it returns usernames, the caller of the method will still have to execute a query to ensure that the usernames are valid. Therefore, I'm failing to understand why we can't return `Post#mentions_user_ids` instead. 

Also I just realized that since we're no longer combining the lookups, we actually end up with a executing way more Redis calls when viewing a topic. I tested out the changes out locally and this increased the number of redis calls when viewing a topic to 53.

Reply:
> I'm not sure if this is necessary. `Post#mentions` returns usernames right now. Even though it returns usernames, the caller of the method will still have to execute a query to ensure that the usernames are valid.

No, it won't have to. It works like this – you can mention a non-existent user and nothing will blow up, and that without making any db query to check usernames:

<img width="801" alt="image" src="https://user-images.githubusercontent.com/1274517/212892707-dcc86d95-d6fd-485b-896e-740b16cbff72.png">

It worked this way before we added status to mentions. Then when I was adding status I made sure it proceeds working this way. And with this PR it also remains the same, you can mention non-existent users and everything works without additional queries to db.

> Therefore, I'm failing to understand why we can't return `Post#mentions_user_ids` instead.

It's here, in `PostSerializer`, sometimes we have `TopicView` and sometimes we have only a post:

https://github.com/discourse/discourse/blob/8c792dc551de95e188b6ab8d7e5d72b4166d4968/app/serializers/post_serializer.rb#L564-L568

If we return `mentions_user_ids` from a post, it means that we need to parse post and load user ids for usernames there:

```ruby
class Post
  def mentions_user_ids
    # check_cache
    # parse_usernames_from_post
    # load_user_ids_for_usernames
    # add_to_cash

    return user_ids
  end
end
```

That's fine if we have only one post, but if we have many of them and want to combine queries, we can't just call `post#mentions_user_ids` from `TopicView`, because that would mean all these requests will be made once for every post in `TopicView`.

So if we want to combine queries, we need to do the job both in the post model and in `TopicView`. I've pushed a commit that's doing that. There is no that much code duplication there, but the new code in `TopicView` is relatively complex. The main reason is that let's say we want to load from cache mentions for 20 posts, it may happen that mentions for a part of those posts were cached before, but mentions for another part weren't, so we need to handle that.

If we agree on this solution, I'll cover it with more tests before merging.

By the way:

> I tested out the changes out locally and this increased the number of redis calls when viewing a topic to 53.

I don't think the amount of cache requests says something here. There are fast and are far from being a hot spot in this particular case.

Here what I see on my machine, after combining requests to Redis:

<img width="415" alt="Screenshot 2023-01-17 at 14 38 43" src="https://user-images.githubusercontent.com/1274517/212896204-c8a359a6-b12a-425f-9daa-7804639cdadf.png">

1,2% agains 2,6% I [measured before](https://github.com/discourse/discourse/pull/19644#issue-1513057535). So we won something around of 1,4% on my machine and probably less in production (I've briefly compared how long an average Redis call takes on my local machine and on a Digital Ocean instance, and it's more than two times faster in DO).

But we're winning some memory by switching from storing usernames to storing ids for sure.

Reply:
> So if we want to combine queries, we need to do the job both in the post model and in TopicView. I've pushed a commit that's doing that. There is no that much code duplication there, but the new code in TopicView is relatively complex. The main reason is that let's say we want to load from cache mentions for 20 posts, it may happen that mentions for a part of those posts were cached before, but mentions for another part weren't, so we need to handle that.

Yes combining the lookups have to be done irrespective of whether we opts to store `User#id` or `User#username` in the cache or not. The only difference when using `User#id` vs `User#username` is that when setting the cache, we need an extra query to fetch the user ids when we only have the usernames. 

> I don't think the amount of cache requests says something here.

If we look at the request individual then it might not seem like much. However, we have to consider that `TopicView` is one of the most viewed route for the application so efficiency is important. I also don't think it is just about the roundtrip overhead as we also have to consider that Redis now has to execute 5-6X more commands each time it loads a batch of posts. At high load, this increases the amount of pressure we're putting on our Redis servers.   

> If we agree on this solution, I'll cover it with more tests before merging.

I personally think the current code is OK though I have some suggestions.